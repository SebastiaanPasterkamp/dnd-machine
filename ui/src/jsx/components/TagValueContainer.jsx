import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

import utils from '../utils.jsx';

import '../../sass/_tag-value-container.scss';

import {
    TagsContainer,
    Tag,
    TagBadge,
} from './BaseTagContainer.jsx';
import LazyComponent from './LazyComponent.jsx';
import SingleSelect from '../components/SingleSelect.jsx';

export class TagValueContainer extends LazyComponent
{
    constructor(props) {
        super(props);
        this.funcCache = {};
    }

    memoize = (key, func) => {
        if (!(key in this.funcCache)) {
            this.funcCache[key] = func;
        }
        return this.funcCache[key];
    }

    onChange(key, newValue) {
        const { value, onChange, setState } = this.props;
        setState(
            _.pickBy(
                _.assign(
                    {},
                    value,
                    {[key]: newValue}
                ),
                value => (value !== undefined)
            )
        );

        if (onChange) {
            onChange(key, newValue);
        }
    }

    onDelete(key) {
        const { value, onDelete, setState } = this.props;

        setState(
            _.pickBy(
                _.assign(
                    {},
                    value,
                    {[key]: undefined}
                ),
                value => (value !== undefined)
            )
        );

        if (onDelete) {
            onDelete(key);
        }
    }

    onAdd(key) {
        const {
            value, defaultValue = 0, onAdd, setState,
        } = this.props;

        setState(
            _.pickBy(
                _.assign(
                    {},
                    value,
                    {[key]: defaultValue}
                ),
                value => (value !== undefined)
            )
        );

        if (onAdd) {
            onAdd(key, defaultValue);
        }
    }

    renderSelect() {
        const {
            value, items = [], disabled, showSelect = true,
        } = this.props;

        if (disabled) return null;
        if (!showSelect) return null;
        if (!items.length) return null;

        const filtered = _.chain(items)
            .filter(item => !(
                _.get(item, 'code', item.name) in value
            ))
            .map(item => _.pickBy(
                {
                    code: _.get(item, 'code', item.name),
                    label: _.get(item, 'label', item.name),
                    description: item.description,
                },
                v => v !== undefined
            ))
            .value();

        if (!filtered.length) return null;

        return <SingleSelect
            emptyLabel="Add..."
            items={filtered}
            setState={item => this.onAdd(item)}
            />;
    }

    renderBadge(badgeItems, tag) {
        const { disabled = false, } = this.props;
        const label = _.get(
            _.find(badgeItems || {}, {key: tag.value}),
            'label',
            tag.value
        );

        if (disabled) {
            return <TagBadge>
                { label }
            </TagBadge>;
        }

        if (!badgeItems) {
            return <TagBadge>
                <input
                    type="number"
                    value={tag.value}
                    disabled={disabled}
                    onChange={e => this.onChange(
                        tag.key,
                        e.target.value
                    )}
                    />
            </TagBadge>;
        }

        return <TagBadge>
            { label }
            <div className="nice-tag-dropdown hover">
                <ul>
                {_.map(badgeItems, option => (
                    <li
                        key={option.key}
                        data-value={option.key}
                        onClick={this.memoize(
                            `onClick-${tag.key}`,
                            () => this.onChange(tag.key, option.key)
                        )}
                        >
                        <a className="cursor-pointer">
                            { option.label }
                        </a>
                    </li>
                ))}
                </ul>
            </div>
        </TagBadge>;
    }

    render() {
        const {
            value, tagValues, items = [], setState, className,
            disabled,
        } = this.props;
        const style = utils.makeStyle({}, [
            "tag-value-container",
            className,
        ]);

        const badgeItems = tagValues
            ? _.map(tagValues, item => ({
                key: _.get(item, 'code', item.name),
                label: _.get(item, 'label', item.name),
            }))
            : null;

        const tags = _.map(value, (value, key) => {
            const { label, description } = (
                _.find(items, { code: key })
                || _.find(items, { name: key })
                || {}
            );

            return {
                key,
                value,
                label,
                description,
                disabled,
                onChange: this.memoize(
                    `onChange-${key}`,
                    (value) => this.onChange(key, value)
                ),
                onDelete: this.memoize(
                    `onDelete-${key}`,
                    () => this.onDelete(key)
                ),
            };
        });

        return <TagsContainer className={ style }>
            {this.renderSelect()}

            {_.map(tags, tag => (
                <Tag {...tag}>
                    {this.renderBadge(badgeItems, tag)}
                </Tag>
            ))}
        </TagsContainer>;
    }
}

TagValueContainer.propTypes = {
    value: PropTypes.object.isRequired,
    tagValues: PropTypes.arrayOf(
        PropTypes.object
    ),
    className: PropTypes.string,
    showSelect: PropTypes.bool,
    disabled: PropTypes.bool,
    items: PropTypes.arrayOf(
        PropTypes.object
    ),
    onAdd: PropTypes.func,
    onChange: PropTypes.func,
    onDelete: PropTypes.func,
};

export default TagValueContainer;
