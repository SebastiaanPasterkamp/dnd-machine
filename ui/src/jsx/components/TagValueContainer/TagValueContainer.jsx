import React from 'react';
import PropTypes from 'prop-types';
import {
    find,
    keys,
    map,
} from 'lodash/fp';

import utils, { memoize } from '../../utils';

import './sass/_tag-value-container.scss';

import {
    BaseTagContainer,
    Tag,
    TagBadge,
} from '../BaseTagContainer';
import SingleSelect from '../SingleSelect';

import TagSelect from './components/TagSelect';

export class TagValueContainer extends React.Component
{
    constructor(props) {
        super(props);
        this.memoize = memoize.bind(this);
        this.onAdd = this.onAdd.bind(this);
    }

    onAdd(key) {
        const { value, defaultValue, onAdd, setState } = this.props;

        setState({
            ...value,
            [key]: defaultValue,
        });

        if (onAdd) {
            onAdd(key, defaultValue);
        }
    }

    onChange(key) {
        return this.memoize(
            `onChange-${key}`,
            e => {
                const { value, onChange, setState } = this.props;
                setState({
                    ...value,
                    [key]: e.target.value,
                });

                if (onChange) {
                    onChange(key, e.target.value);
                }
            }
        );
    }

    onDelete(key) {
        return this.memoize(
            `onDelete-${key}`,
            () => {
                const {
                    value: {
                        [key]: removed,
                        ...remaining,
                    },
                    onDelete,
                    setState,
                } = this.props;
                setState(remaining);

                if (onDelete) {
                    onDelete(key);
                }
            }
        );
    }

    render() {
        const { value, items, className, disabled, showSelect } = this.props;
        const style = utils.makeStyle({}, [
            "tag-value-container",
            className,
        ]);

        return (
            <BaseTagContainer className={ style }>
                {!disabled && showSelect ? (
                    <TagSelect
                        onSelect={this.onAdd}
                        items={items}
                        current={value}
                    />
                ) : null}

                {map(key => {
                    const { name, description } = (
                        find({ id: key }, items)
                        || find({ name: key }, items)
                        || {}
                    );

                    return (
                        <Tag
                            key={key}
                            label={name}
                            description={description}
                            onDelete={this.onDelete(key)}
                            disabled={disabled}
                        >
                            <TagBadge>
                                {disabled ? (
                                    value[key]
                                ) : (
                                    <input
                                        type="number"
                                        value={value[key] || ''}
                                        onChange={this.onChange(key)}
                                    />
                                )}
                            </TagBadge>
                        </Tag>
                    );
                })(keys(value))}
            </BaseTagContainer>
        );
    }
}

TagValueContainer.propTypes = {
    value: PropTypes.objectOf(
        PropTypes.number
    ).isRequired,
    defaultValue: PropTypes.number,
    className: PropTypes.string,
    showSelect: PropTypes.bool,
    disabled: PropTypes.bool,
    items: PropTypes.arrayOf(
        PropTypes.object
    ),
    onAdd: PropTypes.func,
    onChange: PropTypes.func,
    onDelete: PropTypes.func,
};

TagValueContainer.defaultProps = {
    defaultValue: 0,
    className: null,
    showSelect: true,
    disabled: false,
    items: [],
    onAdd: null,
    onChange: null,
    onDelete: null,
};

export default TagValueContainer;
