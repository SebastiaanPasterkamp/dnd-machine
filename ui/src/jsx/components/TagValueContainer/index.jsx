export { default, TagValueContainer } from './TagValueContainer';
