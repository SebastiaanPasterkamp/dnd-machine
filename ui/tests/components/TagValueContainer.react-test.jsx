import React from 'react';
import { mount } from 'enzyme';
import TagValueContainer from 'components/TagValueContainer.jsx';
import renderer from 'react-test-renderer';

const items = [
    { code: "a", label: "Aaa" },
    { name: "b", label: "Bbb", description: "Bee" },
];
const tagValues = [
    { code: 0, label: 'Zero' },
    { name: 5, label: 'Five' },
];

describe('Component: TagValueContainer', () => {
    it('should render with minimal props', () => {
        const setState = jest.fn();
        const tree = renderer.create(
            <TagValueContainer
                setState={setState}
                value={{}}
                />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render with all props', () => {
        const setState = jest.fn();
        const tree = renderer.create(
            <TagValueContainer
                setState={setState}
                items={items}
                tagValues={tagValues}
                value={{a: 5, b: 0}}
                className="info"
                />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render disabled', () => {
        const setState = jest.fn();
        const tree = renderer.create(
            <TagValueContainer
                setState={setState}
                items={items}
                tagValues={tagValues}
                value={{a: 5, b: 0}}
                disabled={true}
                />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should exclude selected items from select in single mode', () => {
        const setState = jest.fn();
        const tree = renderer.create(
            <TagValueContainer
                setState={setState}
                items={items}
                value={{a: 5, b: 0}}
                />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should emit changes when adding', () => {
        const setState = jest.fn();
        const onAdd = jest.fn();
        const wrapper = mount(
            <TagValueContainer
                setState={setState}
                onAdd={onAdd}
                items={items}
                value={{a: 5}}
                multiple={true}
                />
        );

        wrapper
            .find('button.nice-btn')
            .simulate('click');
        wrapper
            .find('li[data-value="b"] a')
            .simulate('click');

        expect(onAdd)
            .toBeCalledWith('b', 0);
        expect(setState)
            .toBeCalledWith({a: 5, b: 0});
    });

    it('should emit changes when changing by dropdown', () => {
        const setState = jest.fn();
        const onChange = jest.fn();
        const wrapper = mount(
            <TagValueContainer
                setState={setState}
                onChange={onChange}
                tagValues={tagValues}
                items={items}
                value={{a: 5}}
                />
        );

        wrapper
            .find('[data-value=0]')
            .simulate('click');

        expect(onChange)
            .toBeCalledWith('a', 0);
        expect(setState)
            .toBeCalledWith({a: 0});
    });

    it('should emit changes when changing input', () => {
        const setState = jest.fn();
        const onChange = jest.fn();
        const wrapper = mount(
            <TagValueContainer
                setState={setState}
                onChange={onChange}
                items={items}
                value={{a: 5, b: 0}}
                />
        );

        wrapper
            .find('input')
            .at(1)
            .simulate('change', { target: {value: 3} });

        expect(onChange)
            .toBeCalledWith('b', 3);
        expect(setState)
            .toBeCalledWith({a: 5, b: 3});
    });

    it('should emit changes when deleting', () => {
        const setState = jest.fn();
        const onDelete = jest.fn();
        const wrapper = mount(
            <TagValueContainer
                setState={setState}
                onDelete={onDelete}
                items={items}
                value={{a: 5, b: 0}}
                />
        );

        wrapper
            .find('button.nice-tag-btn')
            .at(0)
            .simulate('click');

        expect(onDelete)
            .toBeCalledWith('a');
        expect(setState)
            .toBeCalledWith({b: 0});
    });
});
