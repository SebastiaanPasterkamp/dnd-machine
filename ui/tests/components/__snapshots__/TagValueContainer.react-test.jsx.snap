// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: TagValueContainer should exclude selected items from select in single mode 1`] = `
<div
  className="base-tag-container nice-tags-container tag-value-container"
>
  <div
    className="nice-tag"
  >
    <span
      className="nice-tag-label"
    >
      Aaa
    </span>
    <span
      className="nice-tag-badge"
    >
      <input
        disabled={false}
        onChange={[Function]}
        type="number"
        value={5}
      />
    </span>
    <button
      className="nice-tag-btn"
      onClick={[Function]}
    >
      <i
        className="icon fa-trash-o"
      />
    </button>
  </div>
  <div
    className="nice-tag"
  >
    <span
      className="nice-tag-label"
    >
      <div
        className="tool-tip inline"
      >
        Bbb
        <span
          className="tool-tip--content"
        >
          <p>
            Bee
          </p>
        </span>
      </div>
    </span>
    <span
      className="nice-tag-badge"
    >
      <input
        disabled={false}
        onChange={[Function]}
        type="number"
        value={0}
      />
    </span>
    <button
      className="nice-tag-btn"
      onClick={[Function]}
    >
      <i
        className="icon fa-trash-o"
      />
    </button>
  </div>
</div>
`;

exports[`Component: TagValueContainer should render with all props 1`] = `
<div
  className="base-tag-container nice-tags-container tag-value-container"
>
  <div
    className="nice-tag muted"
  >
    <span
      className="nice-tag-label"
    >
      Aaa
    </span>
    <span
      className="nice-tag-badge"
    >
      <div
        className="nice-tag-dropdown hover"
      >
        Five
        <ul>
          <li
            data-value={0}
          >
            <a
              onClick={[Function]}
            >
              Zero
            </a>
          </li>
          <li
            data-value={5}
          >
            <a
              onClick={[Function]}
            >
              Five
            </a>
          </li>
        </ul>
      </div>
    </span>
  </div>
  <div
    className="nice-tag muted"
  >
    <span
      className="nice-tag-label"
    >
      <div
        className="tool-tip inline"
      >
        Bbb
        <span
          className="tool-tip--content"
        >
          <p>
            Bee
          </p>
        </span>
      </div>
    </span>
    <span
      className="nice-tag-badge"
    >
      <div
        className="nice-tag-dropdown hover"
      >
        Zero
        <ul>
          <li
            data-value={0}
          >
            <a
              onClick={[Function]}
            >
              Zero
            </a>
          </li>
          <li
            data-value={5}
          >
            <a
              onClick={[Function]}
            >
              Five
            </a>
          </li>
        </ul>
      </div>
    </span>
  </div>
</div>
`;

exports[`Component: TagValueContainer should render with minimal props 1`] = `
<div
  className="base-tag-container nice-tags-container tag-value-container"
/>
`;
